lms/djangoapps/courseware/model_data.py

```diff
@@ -34,12 +35,473 @@ class InvalidWriteError(Exception):
+
+        Arguments:
+            kvs_key (`DjangoKeyValueStore.Key`): The field value to delete
+            value: The field value to store
+        """
+        self.set_many({kvs_key: value})
+
+    @contract(kvs_key=DjangoKeyValueStore.Key, returns="datetime|None")
+    def last_modified(self, kvs_key):
+        """
+        Return when the supplied field was changed.
+
+        Arguments:
+            kvs_key (`DjangoKeyValueStore.Key`): The field value to delete
+
+        Returns: datetime if there was a modified date, or None otherwise
+        """
+        field_object = self._cache.get(self._cache_key_for_kvs_key(kvs_key))
+
+        if field_object is None:
+            return None
+        else:
+            return field_object.modified
+
+    @contract(kv_dict="dict(DjangoKeyValueStore_Key: *)")
+    def set_many(self, kv_dict):
+        """
+        Set the specified fields to the supplied values.
+
+        Arguments:
+            kv_dict (dict): A dictionary mapping :class:`~DjangoKeyValueStore.Key`
+                objects to values to set.
+        """
+        pending_updates = defaultdict(dict)
+        for kvs_key, value in kv_dict.items():
+            cache_key = self._cache_key_for_kvs_key(kvs_key)
+            field_state = self._cache[cache_key]
+            field_state[kvs_key.field_name] = value
+
+            pending_updates[cache_key][kvs_key.field_name] = value
+
+        try:
+            self._client.set_many(
+                self.user.username,
+                pending_updates
+            )
+        except DatabaseError:
+            raise KeyValueMultiSaveError(sum((state.keys() for state in pending_updates.values()), []))
+
+    @contract(kvs_key=DjangoKeyValueStore.Key)
+    def get(self, kvs_key):
+        """
+        Return the django model object specified by `kvs_key` from
+        the cache.
+
+        Arguments:
+            kvs_key (`DjangoKeyValueStore.Key`): The field value to delete
+
+        Returns: A django orm object from the cache
+        """
+        cache_key = self._cache_key_for_kvs_key(kvs_key)
+        if cache_key not in self._cache:
+            raise KeyError(kvs_key.field_name)
+
+        return self._cache[cache_key][kvs_key.field_name]
+
+    @contract(kvs_key=DjangoKeyValueStore.Key)
+    def delete(self, kvs_key):
+        """
+        Delete the value specified by `kvs_key`.
+
+        Arguments:
+            kvs_key (`DjangoKeyValueStore.Key`): The field value to delete
+
+        Raises: KeyError if key isn't found in the cache
+        """
+        cache_key = self._cache_key_for_kvs_key(kvs_key)
+        if cache_key not in self._cache:
+            raise KeyError(kvs_key.field_name)
+
+        field_state = self._cache[cache_key]
+
+        # If field_name isn't in field_state,
+        # then this will throw the correct KeyError
+        del field_state[kvs_key.field_name]
+
+        self._client.delete(self.user.username, cache_key, fields=[kvs_key.field_name])
```

Comment:
Moved so that the local cache is updated only if the call to the client completes successfully.